codes/kotlin/chapter_backtracking/preorder_traversal_ii_compact.kt

```diff
@@ -10,7 +10,7 @@ import utils.TreeNode
 import utils.printTree
 
 var path: MutableList<TreeNode>? = null
-var res: MutableList<List<TreeNode>>? = null
+var res: ArrayList<List<TreeNode>>? = null
```

Comment:
I am confused with the difference among `List` `ArrayList` `MutableList`. Why do you use different types for the different variables.

Reply:
Sure